packages/core/src/storages/request_provider.ts

```diff
@@ -567,6 +567,30 @@ export abstract class RequestProvider implements IStorage {
      * the function might occasionally return a false negative,
      * but it will never return a false positive.
      */
+    /**
+     * TODO Kuba:
+     * I would move this to request_queue_v2.ts file as it is related to the new API.
+     * Let's clean it up. We would probably need some new API endpoints to make this work.
```

Comment:
> ad. 1. This is little bit new concept, I think if we introduce this one the listAndLockHead will not make sense. Maybe we can add some parameter instead. But can you iterate over how this will help?

I would love to make it _**trivial**_ to use the API from Crawlee. If I could just call this endpoint repeatedly and it would a) return my current locked requests, b) lock more requests if possible and if I have less than `itemCount` and c) prolong locks on those I already have, that would be a step in that direction.

On top of that, if adding requests to the queue also returned the new "queue head", we could get rid of the whole special treatment for forefronted requests.

> ad. 2. This make sense to offload to API, I just do not want to offload the whole logic, as the scraper need to decide if want to finish or not. Plus it has better idea if need to enqueue new URLs or not. Maybe just isHeadEmpty will be ok to have as API>

Which part of the logic would you like to keep in the crawler? Just because `RequestQueue.isFinished()` says that I can finish doesn't mean that e.g. `BasicCrawler` needs to obey it without question if e.g., `keepAlive: true` is used.

Reply:
ad. 1 Yeah, your points make sense, on the other hand, the request queue will be more expensive, because you will need to do more queries to dynamo, plus you need to transfer more data between client and server. I will try to find out how cost will change and if it is doable with dynamo. 

ad 2.
On the backend you have no idea what is happening in crawlee, there can be still some parsing of the site map or so. Maybe we are talking about the same function, but we do not agree on naming, because you can easily say that the queue is empty on backend but you cannot for sure tell that requestQueue is finish, that was my point.

Reply:
> ad. 1 Yeah, your points make sense, on the other hand, the request queue will be more expensive

Well, if it's cheap but nobody actually knows how to use it, that's a lose for us. Bugs and leaky abstractions have already cost us plenty and I've seen users running stuff on Apify with their homebaked queues, which also isn't great.

> ad 2. On the backend you have no idea what is happening in crawlee, there can be still some parsing of the site map or so. Maybe we are talking about the same function, but we do not agree on naming, because you can easily say that the queue is empty on backend but you cannot for sure tell that requestQueue is finish, that was my point.

True, but if someone wants to 1. run crawlee in parallel with a single shared request queue and 2. break the paradigm that you add requests when handling another request from the same queue, I'd argue that they can either just wait for some time (the current grace period) or implement their own synchronization in a way that they see fit.

Reply:
> ad 2. On the backend you have no idea what is happening in crawlee

Also in crawlee you have no idea what's happening in another crawlee process that consumes the same queue :slightly_smiling_face:

Reply:
based on meeting:
ad 1.
maybe implement **ensure queue head** - based if it is possible with current index in dynamo

ad 2.
- document modifiedAt for API Get queue, plus make it from requestQueueCache.
- Into listAndLockHead add flag if there are any locks in queue.